src/OrbitGl/Viewport.cpp

```diff
@@ -0,0 +1,130 @@
+// Copyright (c) 2021 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "Viewport.h"
+
+#include <algorithm>
+
+#include "OrbitBase/Logging.h"
+
+namespace orbit_gl {
+
+Viewport::Viewport(int width, int height) : width_(width), height_(height) {
+  world_width_ = width;
+  world_height_ = height;
+  world_extents_ = Vec2(width, height);
+}
+
+void Viewport::Resize(int width, int height) {
+  CHECK(width > 0);
+  CHECK(height > 0);
+
+  if (width == width_ && height == height_) return;
+
+  width_ = width;
+  height_ = height;
+  FlagAsDirty();
+}
+
+int Viewport::GetWidth() const { return width_; }
+
+int Viewport::GetHeight() const { return height_; }
+
+void Viewport::SetWorldWidth(float width) {
+  if (width == world_width_) return;
+
+  world_width_ = width;
+  // Recalculate required scrolling
+  SetWorldTopLeftX(world_top_left_[0]);
+  FlagAsDirty();
+}
+
+float Viewport::GetWorldWidth() const { return world_width_; }
+
+void Viewport::SetWorldHeight(float height) {
+  if (height == world_height_) return;
+
+  world_height_ = height;
+  // Recalculate required scrolling
+  SetWorldTopLeftY(world_top_left_[1]);
+
+  FlagAsDirty();
+}
+
+float Viewport::GetWorldHeight() const { return world_height_; }
+
+void Viewport::SetWorldExtents(float width, float height) {
+  Vec2 size = Vec2(width, height);
+
+  if (size == world_extents_) return;
+
+  world_extents_ = size;
+  // Recalculate required scrolling
+  SetWorldTopLeftX(world_top_left_[0]);
+  SetWorldTopLeftY(world_top_left_[1]);
+
+  FlagAsDirty();
+}
+
+Vec2 Viewport::GetWorldExtents() { return world_extents_; }
+
+void Viewport::SetWorldTopLeftY(float y) {
+  float clamped = std::min(std::max(y, world_height_ - world_extents_[1]), 0.f);
```

Comment:
I've decided against this for 2 reasons:
1) I want the user of the class to be able to zoom out to more than 100% - for which you need to set world height > world extents
2) I don't want changes in the visible height to affect the extents because this could lead to unwanted side-effects.

I couldn't use `std::clamp` unfortunately since it doesn't deal with the case `world_extents_[1] < world_height_` correctly.

Reply:
I see your point now (and actually I think I modified it in the past when I made the vertical zoom and thought the same). My plan seems to be clearly more dirty and bug-prone. Thanks for the explanation :)